samples/opengl/depthMapTriangulation/image_triangulator.cpp

```diff
@@ -0,0 +1,949 @@
+/**
+ * Copyright (c) 2020 Simon Mederer
+ * Licensed under the MIT License.
+ * See LICENSE file for more information.
+ */
+
+#include "image_triangulator.h"
+
+// ------------------- naive triangulation -------------------
+
+// ----  PUBLIC  ----
+Saiga::SimpleTriangulator::SimpleTriangulator(Settings const& settings_in) : settings(settings_in) {}
+
+Saiga::SimpleTriangulator::~SimpleTriangulator() {}
+
+void Saiga::SimpleTriangulator::triangulate_image(ImageView<const float> depthImage, OpenTriangleMesh& mesh_out)
+{
+    // create an ImageView to hold the vertexHandles that are added to the mesh
+    std::vector<OpenMesh::VertexHandle> pixel_vertexHandles(depthImage.height * depthImage.width);
+    ImageView<OpenMesh::VertexHandle> pixel_vertexHandles_iV(depthImage.height, depthImage.width,
+                                                             pixel_vertexHandles.data());
+
+    // unproject and add all valid vertices to the mesh
+    add_vertices_to_mesh(depthImage, pixel_vertexHandles_iV, mesh_out);
+
+    // triangulate the added vertices
+    completely_triangulate(mesh_out, pixel_vertexHandles_iV);
+
+
+    // Post-computation
+
+    // activate status so that items can be deleted
+    mesh_out.request_face_status();
+    mesh_out.request_edge_status();
+    mesh_out.request_vertex_status();
+
+    // delete unnecessarily added vertices
+    mesh_out.delete_isolated_vertices();
+    mesh_out.garbage_collection();
+
+    // calculate normals for future use
+    mesh_out.request_face_normals();
+    mesh_out.update_face_normals();
+
+    // release the status stuff
+    mesh_out.release_face_status();
+    mesh_out.release_edge_status();
+    mesh_out.release_vertex_status();
+}
+
+// ----  PRIVATE  ----
+void Saiga::SimpleTriangulator::add_face_to_mesh(MyMesh& mesh, MyMesh::VertexHandle vh1, MyMesh::VertexHandle vh2,
+                                                 MyMesh::VertexHandle vh3)
+{
+    if (mesh.point(vh1)[2] == settings.broken_values || mesh.point(vh2)[2] == settings.broken_values || mesh.point(vh3)[2] == settings.broken_values) return;
+
+    mesh.add_face(vh1, vh2, vh3);
+}
+
+void Saiga::SimpleTriangulator::add_vertices_to_mesh(ImageView<const float> depthImageView,
+                                                     ImageView<OpenMesh::VertexHandle> pixel_vertices, MyMesh& mesh)
+{
+    int height = depthImageView.height;
+    int width  = depthImageView.width;
+
+    assert(pixel_vertices.height == height);
+    assert(pixel_vertices.width == width);
+
+    ImageProcessor::Settings ip_settings;
+    ip_settings.cameraParameters = settings.cameraParameters;
+    ImageProcessor ip(ip_settings);
+
+    std::vector<OpenMesh::Vec3f> unprojected_image(height * width);
+    ImageView<OpenMesh::Vec3f> unprojected_image_iV(height, width, unprojected_image.data());
```

Comment:
`TemplatedImage<Vec3f>`

Reply:
`TemplatedImage` seems to also break with both `OpenMesh::Vec3f` and `MyMesh::VertexHandle`. Would it be better to just omit the ImageView and directly use the vector?

Reply:
This is fine. Keep it like that.